### PR TITLE
Ethercat logging - Larger Jsons

### PR DESCRIPTION
**Pull request by slacAWallace**
Tried B-hoff's suggestion with good results. Jsons can now be up to 7k chars (at least I tested it past 256 and it worked...). @klauer

### CODE REVIEW

**Comment by slacAWallace** [comment]
{
  "ecat_diag": [
    "nIndex",
    1,
    "sName",
    "Term 3 (EK1122)",
    "sType",
    "EK1122-0000-0018",
    "nECAddr",
    1001,
    "bDiagData",
    false,
    "stPortCRCErrors",
    {
      "portA": 0,
      "portB": 0,
      "portC": 0,
      "portD": 0
    },
    "nSumCRCErrors",
    0,
    "stState",
    {
      "eEcState ": 8,
      "nReserved": 0,
      "bError": false,
      "bInvalidVPRS": false,
      "nReserved2": 0,
      "bNoCommToSlave": false,
      "bLinkError": false,
      "bMissingLink": false,
      "bUnexpectedLink": false,
      "bPortA": false,
      "bPortB": false,
      "bPortC": false,
      "bPortD": false
    },
    "nIndex",
    2,
    "sName",
    "SubSysPLC1",
    "sType",
    "EL6692-0002-0020",
    "nECAddr",
    1002,
    "bDiagData",
    true,
    "stPortCRCErrors",
    {
      "portA": 0,
      "portB": 0,
      "portC": 0,
      "portD": 0
    },
    "nSumCRCErrors",
    0,
    "stState",
    {
      "eEcState ": 1,
      "nReserved": 0,
      "bError": false,
      "bInvalidVPRS": false,
      "nReserved2": 0,
      "bNoCommToSlave": true,
      "bLinkError": false,
      "bMissingLink": false,
      "bUnexpectedLink": false,
      "bPortA": false,
      "bPortB": false,
      "bPortC": false,
      "bPortD": false
    },
    "nIndex",
    3,
    "sName",
    "SubSysPLC2",
    "sType",
    "EL6692-0002-0020",
    "nECAddr",
    1003,
    "bDiagData",
    false,
    "stPortCRCErrors",
    {
      "portA": 0,
      "portB": 0,
      "portC": 0,
      "portD": 0
    },
    "nSumCRCErrors",
    0,
    "stState",
    {
      "eEcState ": 8,
      "nReserved": 0,
      "bError": false,
      "bInvalidVPRS": false,
      "nReserved2": 0,
      "bNoCommToSlave": false,
      "bLinkError": false,
      "bMissingLink": false,
      "bUnexpectedLink": false,
      "bPortA": false,
      "bPortB": false,
      "bPortC": false,
      "bPortD": false
    },
    "nIndex",
    4,
    "sName",
    "Term 6 (EK1110)",
    "sType",
    "EK1110-0000-0017",
    "nECAddr",
    1004,
    "bDiagData",
    false,
    "stPortCRCErrors",
    {
      "portA": 0,
      "portB": 0,
      "portC": 0,
      "portD": 0
    },
    "nSumCRCErrors",
    0,
    "stState",
    {
      "eEcState ": 8,
      "nReserved": 0,
      "bError": false,
      "bInvalidVPRS": false,
      "nReserved2": 0,
      "bNoCommToSlave": false,
      "bLinkError": false,
      "bMissingLink": false,
      "bUnexpectedLink": false,
      "bPortA": false,
      "bPortB": false,
      "bPortC": false,
      "bPortD": false
    },
    "nIndex",
    0,
    "sName",
    "",
    "sType",
    "",
    "nECAddr",
    0,
    "bDiagData",
    false,
    "stPortCRCErrors",
    {
      "portA": 0,
      "portB": 0,
      "portC": 0,
      "portD": 0
    },
    "nSumCRCErrors",
    0,
    "stState",
    {
      "eEcState ": 0,
      "nReserved": 0,
      "bError": false,
      "bInvalidVPRS": false,
      "nReserved2": 0,
      "bNoCommToSlave": false,
      "bLinkError": false,
      "bMissingLink": false,
      "bUnexpectedLink": false,
      "bPortA": false,
      "bPortB": false,
      "bPortC": false,
      "bPortD": false
    }
  ]
}

**Comment by slacAWallace** [comment]
{
  "ecat_diag": [
    "nECAddr",
    1001,
    {
      "nIndex": 1,
      "sName": "Term 3 (EK1122)",
      "sType": "EK1122-0000-0018",
      "bDiagData": false,
      "stPortCRCErrors": {
        "portA": 0,
        "portB": 0,
        "portC": 0,
        "portD": 0
      },
      "nSumCRCErrors": 0,
      "stState": {
        "eEcState ": 8,
        "nReserved": 0,
        "bError": false,
        "bInvalidVPRS": false,
        "nReserved2": 0,
        "bNoCommToSlave": false,
        "bLinkError": false,
        "bMissingLink": false,
        "bUnexpectedLink": false,
        "bPortA": false,
        "bPortB": false,
        "bPortC": false,
        "bPortD": false
      }
    },
    "nECAddr",
    1002,
    {
      "nIndex": 2,
      "sName": "SubSysPLC1",
      "sType": "EL6692-0002-0020",
      "bDiagData": true,
      "stPortCRCErrors": {
        "portA": 0,
        "portB": 0,
        "portC": 0,
        "portD": 0
      },
      "nSumCRCErrors": 0,
      "stState": {
        "eEcState ": 1,
        "nReserved": 0,
        "bError": false,
        "bInvalidVPRS": false,
        "nReserved2": 0,
        "bNoCommToSlave": true,
        "bLinkError": false,
        "bMissingLink": false,
        "bUnexpectedLink": false,
        "bPortA": false,
        "bPortB": false,
        "bPortC": false,
        "bPortD": false
      }
    },
    "nECAddr",
    1003,
    {
      "nIndex": 3,
      "sName": "SubSysPLC2",
      "sType": "EL6692-0002-0020",
      "bDiagData": false,
      "stPortCRCErrors": {
        "portA": 0,
        "portB": 0,
        "portC": 0,
        "portD": 0
      },
      "nSumCRCErrors": 0,
      "stState": {
        "eEcState ": 8,
        "nReserved": 0,
        "bError": false,
        "bInvalidVPRS": false,
        "nReserved2": 0,
        "bNoCommToSlave": false,
        "bLinkError": false,
        "bMissingLink": false,
        "bUnexpectedLink": false,
        "bPortA": false,
        "bPortB": false,
        "bPortC": false,
        "bPortD": false
      }
    },
    "nECAddr",
    1004,
    {
      "nIndex": 4,
      "sName": "Term 6 (EK1110)",
      "sType": "EK1110-0000-0017",
      "bDiagData": false,
      "stPortCRCErrors": {
        "portA": 0,
        "portB": 0,
        "portC": 0,
        "portD": 0
      },
      "nSumCRCErrors": 0,
      "stState": {
        "eEcState ": 8,
        "nReserved": 0,
        "bError": false,
        "bInvalidVPRS": false,
        "nReserved2": 0,
        "bNoCommToSlave": false,
        "bLinkError": false,
        "bMissingLink": false,
        "bUnexpectedLink": false,
        "bPortA": false,
        "bPortB": false,
        "bPortC": false,
        "bPortD": false
      }
    },
    "nECAddr",
    0,
    {
      "nIndex": 0,
      "sName": "",
      "sType": "",
      "bDiagData": false,
      "stPortCRCErrors": {
        "portA": 0,
        "portB": 0,
        "portC": 0,
        "portD": 0
      },
      "nSumCRCErrors": 0,
      "stState": {
        "eEcState ": 0,
        "nReserved": 0,
        "bError": false,
        "bInvalidVPRS": false,
        "nReserved2": 0,
        "bNoCommToSlave": false,
        "bLinkError": false,
        "bMissingLink": false,
        "bUnexpectedLink": false,
        "bPortA": false,
        "bPortB": false,
        "bPortC": false,
        "bPortD": false
      }
    }
  ]
}

File: LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU [hunk 1]
```diff
@@ -215,55 +205,57 @@ END_IF
...
-        fbJson.AddKey('ecat_master_bResetActive');
-        fbJson.AddBool(stMasterDevState.bResetActive);
-        fbJson.AddKey('ecat_master_bResetRequired');
-        fbJson.AddBool(stMasterDevState.bResetRequired);
-        fbJson.AddKey('ecat_master_bWatchdogTriggerd');
-        fbJson.AddBool(stMasterDevState.bWatchdogTriggerd);
-        fbJson.AddKey('ecat_master_eEcState');
-        fbJson.AddUdint(stMasterDevState.eEcState);
-        fbJson.EndObject();
+    fbJson.AddKey('ecat_master_bAtLeastOneNotInOp');
```
Comment (klauer) [inline LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU]: One more: `ecast_master_` could be removed and made into a sub-dict:

```
"ecat_master": {
    "bAtLeastOneNotInOp": ...,
    "bDcNotInSync": ...,
    ...
}
```

**Comment by klauer** [comment]
IPython shows a bit of structure, perhaps it's better:
```python
{'ecat_diag': ['nECAddr',
  1001,
  {'nIndex': 1,
   'sName': 'Term 3 (EK1122)',
   'sType': 'EK1122-0000-0018',
   'bDiagData': False,
   'stPortCRCErrors': {'portA': 0, 'portB': 0, 'portC': 0, 'portD': 0},
   'nSumCRCErrors': 0,
   'stState': {'eEcState ': 8,
    'nReserved': 0,
    'bError': False,
    'bInvalidVPRS': False,
    'nReserved2': 0,
    'bNoCommToSlave': False,
    'bLinkError': False,
    'bMissingLink': False,
    'bUnexpectedLink': False,
    'bPortA': False,
    'bPortB': False,
    'bPortC': False,
    'bPortD': False}},
  'nECAddr',
  1002,
  {'nIndex': 2,
   'sName': 'SubSysPLC1',
   'sType': 'EL6692-0002-0020',
   'bDiagData': True,
   'stPortCRCErrors': {'portA': 0, 'portB': 0, 'portC': 0, 'portD': 0},
   'nSumCRCErrors': 0,
   'stState': {'eEcState ': 1,
    'nReserved': 0,
    'bError': False,
    'bInvalidVPRS': False,
    'nReserved2': 0,
    'bNoCommToSlave': True,
    'bLinkError': False,
    'bMissingLink': False,
    'bUnexpectedLink': False,
    'bPortA': False,
    'bPortB': False,
    'bPortC': False,
    'bPortD': False}},
  'nECAddr',
  1003,
  {'nIndex': 3,
   'sName': 'SubSysPLC2',
   'sType': 'EL6692-0002-0020',
   'bDiagData': False,
   'stPortCRCErrors': {'portA': 0, 'portB': 0, 'portC': 0, 'portD': 0},
   'nSumCRCErrors': 0,
   'stState': {'eEcState ': 8,
    'nReserved': 0,
    'bError': False,
    'bInvalidVPRS': False,
    'nReserved2': 0,
    'bNoCommToSlave': False,
    'bLinkError': False,
    'bMissingLink': False,
    'bUnexpectedLink': False,
    'bPortA': False,
    'bPortB': False,
    'bPortC': False,
    'bPortD': False}},
  'nECAddr',
  1004,
  {'nIndex': 4,
   'sName': 'Term 6 (EK1110)',
   'sType': 'EK1110-0000-0017',
   'bDiagData': False,
   'stPortCRCErrors': {'portA': 0, 'portB': 0, 'portC': 0, 'portD': 0},
   'nSumCRCErrors': 0,
   'stState': {'eEcState ': 8,
    'nReserved': 0,
    'bError': False,
    'bInvalidVPRS': False,
    'nReserved2': 0,
    'bNoCommToSlave': False,
    'bLinkError': False,
    'bMissingLink': False,
    'bUnexpectedLink': False,
    'bPortA': False,
    'bPortB': False,
    'bPortC': False,
    'bPortD': False}},
  'nECAddr',
  0,
  {'nIndex': 0,
   'sName': '',
   'sType': '',
   'bDiagData': False,
   'stPortCRCErrors': {'portA': 0, 'portB': 0, 'portC': 0, 'portD': 0},
   'nSumCRCErrors': 0,
   'stState': {'eEcState ': 0,
    'nReserved': 0,
    'bError': False,
    'bInvalidVPRS': False,
    'nReserved2': 0,
    'bNoCommToSlave': False,
    'bLinkError': False,
    'bMissingLink': False,
    'bUnexpectedLink': False,
    'bPortA': False,
    'bPortB': False,
    'bPortC': False,
    'bPortD': False}}]}
```

There's a bit of a strange mix of `[str, int, dict, str, int, dict...]` in there. 
I wonder if we should have a dictionary keyed on `nECAddr` to info for that node?

**Comment by slacAWallace** [comment]
[Json beauty](https://codebeautify.org/jsonviewer/cbf23622)

Tweaked it again since we were nearing the char limit with only three terminals.

**Comment by slacAWallace** [comment]
The latest should have addressed your point @klauer

**Comment by klauer** [comment]
Looks much better!

**Pull request merged**